src/molecules/Quaternion.cpp

```diff
@@ -1,6 +1,8 @@
 #include "Quaternion.h"
 #include <cmath>
 
+#define VECTOR3_ZERO ((std::array<double, 3>){ 0, 0, 0 })
```

Comment:
Maybe there is already an equal definition? This seems to be quite general

Reply:
A class [Vector3.h](https://github.com/ls1mardyn/ls1-mardyn/blob/f3a8ba4ee5ee343f9eed75bc5eb320741c7f89f2/src/bhfmm/utils/Vector3.h#L17) exists, but this type is not used here. There are a lot of variables of type `std::array<double, 3>` in the code of which none is explicitly assigned `{ 0, 0, 0 }`.  
Thus, “consistency” in this regard is not possible without some more involved refactoring, which might be a bit “overkill”.  
Perhaps it would be appropriate to use a constant private class (static) variable (static) `zero_vector` instead to make it more C++ instead of C-like. (?)

Reply:
You are right, we are now again at the point, where all the legacy code of ls1 comes in 😆. There is also a folder `utils` with lots of files in it (probably some are not used anymore...). So yes, maybe it's easier for now to just leave it as it is.
Due to my lacking (proper) IT background, I don't have an opinion on C++ instead of C-like. For me, it's fine as it is.

Reply:
Some context: Vector3.h is in the `bhfmm` (short for Barnes-Hut Fast Multipole Method) folder. I'm not even sure this stuff is included in the compile sources by default so I'd use it with care or move it. But sure both of you are right. More sane utils are always a good idea.
Regarding C++ vs C: Do you like your code debuggable? Yes? Then use C++ style. [`#define` should be avoided in 9.9/10 cases](https://stackoverflow.com/a/14041847/7019073).

Reply:
The more I think about it, the more it seems to me that in this case, it may actually be permissible to use a macro. It is only a shorthand for a typed zero value and is only used (and useful) in this implementation file, and thus it does not make the code hard to read or debug.
Alternatives:
+ Using the contents of the macro makes the code to read and causes duplication (potential for future bugs)
+ Using a local variable in the method causes duplication (potential for future bug)
+ Introducing a `const` member in the header with no other utility makes the code harder to understand and de-localizes the code.

Reply:
At a first glance, I agree that it is *permissible* by your arguments, however, this does not mean that I think it's a good idea because it lowers the threshold for what macros we are ok with ;)
Now, after actually looking into the code, I would disagree even more. The macro you suggest provides the code to instantiate a nameless array filled with zeros. It is then used in two cases:

1. Initialization of arrays:  
   ```c++
   std::array<double, 3> drot VECTOR3_ZERO;
   ```  
   But this can be replaced by
   ```c++
   std::array<double, 3> drot{};
   ```

2. Checks against input arrays:
   ```c++
   if(VECTOR3_ZERO != d) {
   ```
   Here, I see two options. Either use a zero vector that you declared directly before the if or declare another one that is explicitly named as such:
   ```c++
   constexpr std::array<double, 3> vectorZero{};
   if(vectorZero != d) {
   ```
   Or actually create an inline on-the-fly array.
   
   ```c++
   if(std::array<double, 3>{} != d) {
   ```
   Yes, this is not the most beautiful but IMHO clear-enough solution. Throw in a comment if you feel it makes it better :)

Reply:
It's definitely the cleaner solution. The verbosity of the type is probably the greatest issue for me, but a `typedef` wouldn't exactly be a good idea either.